feature/dynamodb/attributevalue/decode.go

```diff
@@ -0,0 +1,748 @@
+}
+
+// UnmarshalList is an alias for Unmarshal func which unmarshals
+// a slice of AttributeValues.
+//
+// The output value provided must be a non-nil pointer
+func UnmarshalList(l []types.AttributeValue, out interface{}) error {
+	return NewDecoder().Decode(&types.AttributeValueMemberL{Value: l}, out)
+}
+
+// UnmarshalListOfMaps is an alias for Unmarshal func which unmarshals a
+// slice of maps of attribute values.
+//
+// This is useful for when you need to unmarshal the Items from a Query API
+// call.
+//
+// The output value provided must be a non-nil pointer
+func UnmarshalListOfMaps(l []map[string]types.AttributeValue, out interface{}) error {
+	items := make([]types.AttributeValue, len(l))
+	for i, m := range l {
+		items[i] = &types.AttributeValueMemberM{Value: m}
+	}
+
+	return UnmarshalList(items, out)
+}
+
+// DecoderOptions is a collection of options to configure how the decoder
+// unmarshalls the value.
+type DecoderOptions struct {
+	// Support other custom struct tag keys, such as `yaml`, `json`, or `toml`.
+	// Note that values provided with a custom TagKey must also be supported
+	// by the (un)marshalers in this package.
+	//
+	// Tag key `dynamodbav` will always be read, but if custom tag key
+	// conflicts with `dynamodbav` the custom tag key value will be used.
+	TagKey string
+
+	// Instructs the decoder to decode AttributeValue Numbers as
+	// Number type instead of float64 when the destination type
+	// is interface{}. Similar to encoding/json.Number
+	UseNumber bool
+}
+
+// A Decoder provides unmarshaling AttributeValues to Go value types.
+type Decoder struct {
+	options DecoderOptions
+}
+
+// NewDecoder creates a new Decoder with default configuration. Use
+// the `opts` functional options to override the default configuration.
+func NewDecoder(optFns ...func(*DecoderOptions)) *Decoder {
+	var options DecoderOptions
+	for _, fn := range optFns {
+		fn(&options)
+	}
+
+	return &Decoder{
+		options: options,
+	}
+}
+
+// Decode will unmarshal an AttributeValue into a Go value type. An error
+// will be return if the decoder is unable to unmarshal the AttributeValue
+// to the provide Go value type.
+//
+// The output value provided must be a non-nil pointer
+func (d *Decoder) Decode(av types.AttributeValue, out interface{}, opts ...func(*Decoder)) error {
+	v := reflect.ValueOf(out)
+	if v.Kind() != reflect.Ptr || v.IsNil() || !v.IsValid() {
+		return &InvalidUnmarshalError{Type: reflect.TypeOf(out)}
+	}
+
+	return d.decode(av, v, tag{})
+}
+
+var stringInterfaceMapType = reflect.TypeOf(map[string]interface{}(nil))
+var byteSliceType = reflect.TypeOf([]byte(nil))
+var byteSliceSliceType = reflect.TypeOf([][]byte(nil))
+var timeType = reflect.TypeOf(time.Time{})
+
+func (d *Decoder) decode(av types.AttributeValue, v reflect.Value, fieldTag tag) error {
+	var u Unmarshaler
+	_, isNull := av.(*types.AttributeValueMemberNULL)
+	if av == nil || isNull {
```

Comment:
If av is not nil and the type is `AttributeValueMemberNULL` should this check that `av.Value == true`?

Reply:
It would be off if it was set to false, but it's the more correct check.

Reply:
We could check if `Value` is false. I'll check if Dynamo looks at this value. I had a feeling it ignored or always set the value to true on its side.  A value of false doesn't really have any clear meaning. (Other than potentially error)